### PR TITLE
Remove unused feature gates from library/ crates

### PR DESCRIPTION
**Pull request by est31**
Removes some unused feature gates from library crates. It's likely not a complete list as I only tested a subset for which it's more likely that it is unused.

### CODE REVIEW

File: library/alloc/src/lib.rs [hunk 1]
```diff
@@ -103,13 +100,10 @@
...
 #![feature(lang_items)]
 #![feature(layout_for_ptr)]
-#![feature(libc)]
-#![feature(map_first_last)]
-#![feature(map_into_keys_values)]
 #![feature(maybe_uninit_ref)]
 #![feature(negative_impls)]
 #![feature(never_type)]
-#![feature(new_uninit)]
+#![cfg_attr(test, feature(new_uninit))]
```
Comment (lcnr) [inline library/alloc/src/lib.rs:106]: I personally feel like putting a feature behind a `cfg` is slightly weird, but apart from that gj here :+1:

Comment (est31) [reply library/alloc/src/lib.rs:106]: It's done so higher up in the file as well. Should I move it up?

Comment (lcnr) [reply library/alloc/src/lib.rs:106]: wouldn't hurt I think, but I also don't think it's necessary :shrug:

**Comment by lcnr** [comment]
r? @lcnr

Feel free to either move this up or just merge this as is

@bors rollup
@bors delegate+

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://dev.azure.com/rust-lang/rust/_build/results?buildId=36713&view=logs&j=1c64c3ff-4c37-56fe-a133-b407354bbfbf&t=d955a4bf-38f0-58ca-c266-c3c17b7d5f88&l=443) 

@bors r-

**Comment by est31** [comment]
Huh, interesting. The only x86_64 target where `base.max_atomic_width` is set to `Some(128)` is the apple family of targets. That's why I didn't catch it. Does this make sense? It's for a separate issue/PR to figure that out though.

**Comment by est31** [comment]
```
$ rg max_atomic_width | rg 128
compiler/rustc_target/src/spec/aarch64_linux_android.rs:    base.max_atomic_width = Some(128);
compiler/rustc_target/src/spec/aarch64_unknown_none.rs:        max_atomic_width: Some(128),
compiler/rustc_target/src/spec/aarch64_unknown_linux_musl.rs:    base.max_atomic_width = Some(128);
compiler/rustc_target/src/spec/aarch64_apple_ios.rs:            max_atomic_width: Some(128),
compiler/rustc_target/src/spec/aarch64_unknown_netbsd.rs:    base.max_atomic_width = Some(128);
compiler/rustc_target/src/spec/aarch64_apple_darwin.rs:    base.max_atomic_width = Some(128);
compiler/rustc_target/src/spec/x86_64_apple_darwin.rs:    base.max_atomic_width = Some(128); // core2 support cmpxchg16b
compiler/rustc_target/src/spec/aarch64_unknown_none_softfloat.rs:        max_atomic_width: Some(128),
compiler/rustc_target/src/spec/aarch64_unknown_freebsd.rs:    base.max_atomic_width = Some(128);
compiler/rustc_target/src/spec/aarch64_unknown_linux_gnu.rs:    base.max_atomic_width = Some(128);
compiler/rustc_target/src/spec/aarch64_apple_tvos.rs:            max_atomic_width: Some(128),
compiler/rustc_target/src/spec/aarch64_unknown_hermit.rs:    base.max_atomic_width = Some(128);
compiler/rustc_target/src/spec/aarch64_unknown_cloudabi.rs:    base.max_atomic_width = Some(128);
compiler/rustc_target/src/spec/aarch64_wrs_vxworks.rs:    base.max_atomic_width = Some(128);
compiler/rustc_target/src/spec/aarch64_unknown_openbsd.rs:    base.max_atomic_width = Some(128);
compiler/rustc_target/src/spec/aarch64_unknown_redox.rs:    base.max_atomic_width = Some(128);
compiler/rustc_target/src/spec/aarch64_fuchsia.rs:    base.max_atomic_width = Some(128);
```

All aarch64 platforms except for `86_64_apple_darwin.rs`. Does an instruction simply disappear on other OSs? Does apple ship custom microcode? :p.

**Comment by est31** [comment]
The issue needs fixing one way or another. It's been fixed by re-adding the feature gate. re-r? @lcnr

**Comment by est31** [comment]
Regarding my question, see #76952

**Pull request merged**